File: leetcode/dp/Climbing_Stairs/java/SolutionRecursion.java
// https://leetcode.com/problems/climbing-stairs/description/
// 70. Climbing Stairs

// Recursive solution | Top-Down
// Time Complexity: O(2^n) because we are making two recursive calls for each step
// Space Complexity: O(n) because we are using the call stack to store the recursive calls
class SolutionRecursion {
    public int climbStairs(int n) {
        // bases cases
        // one stair can be climbed in only one way
        if (n == 1) return 1;
        // two stairs can be climbed in two ways
        // 1 + 1 steps | 2 steps at once
        if (n == 2) return 2;
        return climbStairs(n-2) + climbStairs(n-1);
    }
    public static void main(String[] args) {
        int n = 45;
        System.out.println("Solution Recursion");
        SolutionRecursion solrecur = new SolutionRecursion();
        System.out.println(solrecur.climbStairs(n));
    }
}

File: leetcode/dp/Climbing_Stairs/python/SolutionRecursion.py
# https://leetcode.com/problems/climbing-stairs/
# 70. Climbing Stairs

# Recursive Solution | Top-Down
# Time Complexity: O(2^n) because we are making two recursive calls for each step
# Space Complexity: O(n) because we are using a recursive stack to store the results of the subproblems
class Solution:
    def climbStairs(self, n: int) -> int:
        if n == 1:
            return 1
        if n == 2:
            return 2
        return self.climbStairs(n - 1) + self.climbStairs(n - 2)

if __name__ == "__main__":
    n = 45
    expected = 1836311903
    solution = Solution()
    actual = solution.climbStairs(n)
    if actual == expected:
        print("Test Passed")
    else:
        print(f"Test Failed: Expected {expected}, Actual {actual}")